File: mobile/lib/services/challenge_service.dart
import 'package:flutter/foundation.dart';
import '../models/challenge.dart';
import 'api_client.dart';

class ChallengeService extends ChangeNotifier {
  final ApiClient _client;

  ChallengeService(this._client);

  List<Challenge> _challenges = [];
  List<ActiveChallenge> _activeChallenges = [];
  bool _isLoading = false;
  String? _error;

  List<Challenge> get challenges => _challenges;
  List<ActiveChallenge> get activeChallenges => _activeChallenges;
  bool get isLoading => _isLoading;
  String? get error => _error;

  // Fetch all available challenges
  Future<void> fetchChallenges() async {
    _isLoading = true;
    _error = null;
    notifyListeners();

    try {
      final json = await _client.get('/challenges');
      final data = json['data'] as List;
      _challenges = data.map((c) => Challenge.fromJson(c as Map<String, dynamic>)).toList();
    } on ApiException catch (e) {
      _error = e.message;
    } catch (e) {
      _error = 'Failed to load challenges';
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  // Fetch challenge details with rituals
  Future<ChallengeDetails?> getChallengeDetails(int challengeId) async {
    try {
      final json = await _client.get('/challenges/$challengeId');
      final data = json['data'] as Map<String, dynamic>;
      
      final challenge = Challenge.fromJson(data['challenge'] as Map<String, dynamic>);
      final rituals = (data['rituals'] as List)
          .map((r) => Ritual.fromJson(r as Map<String, dynamic>))
          .toList();
      
      UserProgress? progress;
      if (data['userProgress'] != null) {
        progress = UserProgress.fromJson(data['userProgress'] as Map<String, dynamic>);
      }

      return ChallengeDetails(
        challenge: challenge,
        rituals: rituals,
        userProgress: progress,
      );
    } on ApiException catch (e) {
      _error = e.message;
      return null;
    }
  }

  // Fetch active challenges for the user
  Future<void> fetchActiveChallenges() async {
    try {
      final json = await _client.get('/challenges/my/active');
      final data = json['data'] as List;
      _activeChallenges = data.map((ac) => ActiveChallenge.fromJson(ac as Map<String, dynamic>)).toList();
      notifyListeners();
    } on ApiException catch (e) {
      _error = e.message;
      notifyListeners();
    }
  }

  // Start a challenge
  Future<bool> startChallenge(int challengeId) async {
    try {
      await _client.post('/challenges/$challengeId/start');
      await fetchActiveChallenges(); // Refresh active challenges
      return true;
    } on ApiException catch (e) {
      _error = e.message;
      notifyListeners();
      return false;
    }
  }

  // Pause a challenge
  Future<bool> pauseChallenge(int challengeId) async {
    try {
      await _client.post('/challenges/$challengeId/pause');
      await fetchActiveChallenges(); // Refresh active challenges
      return true;
    } on ApiException catch (e) {
      _error = e.message;
      notifyListeners();
      return false;
    }
  }

  // Resume a challenge
  Future<bool> resumeChallenge(int challengeId) async {
    try {
      await _client.post('/challenges/$challengeId/resume');
      await fetchActiveChallenges(); // Refresh active challenges
      return true;
    } on ApiException catch (e) {
      _error = e.message;
      notifyListeners();
      return false;
    }
  }

  // Complete a ritual
  Future<bool> completeRitual(int challengeId, int ritualId) async {
    try {
      await _client.post('/challenges/$challengeId/rituals/$ritualId/complete');
      await fetchActiveChallenges(); // Refresh active challenges
      return true;
    } on ApiException catch (e) {
      _error = e.message;
      notifyListeners();
      return false;
    }
  }

  // Get user's progress in a challenge
  Future<ChallengeProgress?> getProgress(int challengeId) async {
    try {
      final json = await _client.get('/challenges/$challengeId/progress');
      final data = json['data'] as Map<String, dynamic>;
      return ChallengeProgress.fromJson(data);
    } on ApiException catch (e) {
      _error = e.message;
      return null;
    }
  }

  void clearError() {
    _error = null;
    notifyListeners();
  }
}

// Data classes for complex responses
class ChallengeDetails {
  final Challenge challenge;
  final List<Ritual> rituals;
  final UserProgress? userProgress;

  ChallengeDetails({
    required this.challenge,
    required this.rituals,
    this.userProgress,
  });
}

class UserProgress {
  final UserChallenge userChallenge;
  final List<int> completedRitualIds;

  UserProgress({
    required this.userChallenge,
    required this.completedRitualIds,
  });

  factory UserProgress.fromJson(Map<String, dynamic> json) {
    return UserProgress(
      userChallenge: UserChallenge.fromJson(json['userChallenge'] as Map<String, dynamic>),
      completedRitualIds: List<int>.from(json['completedRitualIds'] ?? []),
    );
  }
}

class ActiveChallenge {
  final int id;
  final int userId;
  final int challengeId;
  final ChallengeStatus status;
  final int currentDay;
  final DateTime? startedAt;
  final DateTime? pausedAt;
  final Challenge challenge;
  final Ritual? todayRitual;
  final List<int> completedRitualIds;

  ActiveChallenge({
    required this.id,
    required this.userId,
    required this.challengeId,
    required this.status,
    required this.currentDay,
    this.startedAt,
    this.pausedAt,
    required this.challenge,
    this.todayRitual,
    required this.completedRitualIds,
  });

  factory ActiveChallenge.fromJson(Map<String, dynamic> json) {
    return ActiveChallenge(
      id: json['id'] as int,
      userId: json['userId'] as int,
      challengeId: json['challengeId'] as int,
      status: UserChallenge.parseStatus(json['status'] as String),
      currentDay: json['currentDay'] as int,
      startedAt: json['startedAt'] != null ? DateTime.parse(json['startedAt'] as String) : null,
      pausedAt: json['pausedAt'] != null ? DateTime.parse(json['pausedAt'] as String) : null,
      challenge: Challenge.fromJson(json['challenge'] as Map<String, dynamic>),
      todayRitual: json['todayRitual'] != null 
          ? Ritual.fromJson(json['todayRitual'] as Map<String, dynamic>)
          : null,
      completedRitualIds: List<int>.from(json['completedRitualIds'] ?? []),
    );
  }
}

class ChallengeProgress {
  final UserChallenge userChallenge;
  final Challenge challenge;
  final List<Ritual> rituals;
  final List<int> completedRitualIds;

  ChallengeProgress({
    required this.userChallenge,
    required this.challenge,
    required this.rituals,
    required this.completedRitualIds,
  });

  factory ChallengeProgress.fromJson(Map<String, dynamic> json) {
    return ChallengeProgress(
      userChallenge: UserChallenge.fromJson(json as Map<String, dynamic>),
      challenge: Challenge.fromJson(json['challenge'] as Map<String, dynamic>),
      rituals: (json['rituals'] as List).map((r) => Ritual.fromJson(r as Map<String, dynamic>)).toList(),
      completedRitualIds: List<int>.from(json['completedRitualIds'] ?? []),
    );
  }
}

File: api/src/modules/challenge/challenge.service.ts
import prisma from '../../lib/prismaClient';
import { AppError } from '../../middleware/errorHandler';
import { ChallengeStatus } from '@prisma/client';

class ChallengeService {
  /**
   * Get all available challenges
   */
  async getChallenges() {
    const challenges = await prisma.challenge.findMany({
      where: {
        isActive: true,
      },
      orderBy: {
        sortOrder: 'asc',
      },
      select: {
        id: true,
        title: true,
        subtitle: true,
        description: true,
        duration: true,
        category: true,
        goals: true,
        icon: true,
        colorTheme: true,
        createdAt: true,
        _count: {
          select: {
            rituals: true,
          },
        },
      },
    });

    return challenges;
  }

  /**
   * Get challenge details with all rituals
   */
  async getChallengeDetails(challengeId: number, userId?: number) {
    const challenge = await prisma.challenge.findUnique({
      where: {
        id: challengeId,
        isActive: true,
      },
      include: {
        rituals: {
          where: {
            isActive: true,
          },
          orderBy: {
            dayNumber: 'asc',
          },
        },
      },
    });

    if (!challenge) {
      throw new AppError('Challenge not found', 404);
    }

    // Get user's progress if authenticated
    let userProgress = null;
    if (userId) {
      userProgress = await prisma.userChallenge.findUnique({
        where: {
          userId_challengeId: {
            userId,
            challengeId,
          },
        },
        include: {
          completions: {
            select: {
              ritualId: true,
              completedAt: true,
            },
          },
        },
      });
    }

    return {
      challenge,
      userProgress,
    };
  }

  /**
   * Get user's active challenges
   */
  async getActiveChallenges(userId: number) {
    const userChallenges = await prisma.userChallenge.findMany({
      where: {
        userId,
        status: {
          in: ['ACTIVE', 'PAUSED'],
        },
      },
      include: {
        challenge: {
          select: {
            id: true,
            title: true,
            subtitle: true,
            description: true,
            duration: true,
            category: true,
            goals: true,
            icon: true,
            colorTheme: true,
          },
        },
        completions: {
          select: {
            ritualId: true,
          },
        },
      },
      orderBy: {
        startedAt: 'desc',
      },
    });

    // Get today's ritual for each active challenge
    const activeChallengesWithRituals = await Promise.all(
      userChallenges.map(async (uc) => {
        const todayRitual = await prisma.ritual.findFirst({
          where: {
            challengeId: uc.challengeId,
            dayNumber: uc.currentDay,
            isActive: true,
          },
        });

        const completedRitualIds = new Set(uc.completions.map(c => c.ritualId));

        return {
          ...uc,
          todayRitual,
          completedRitualIds: Array.from(completedRitualIds),
        };
      })
    );

    return activeChallengesWithRituals;
  }

  /**
   * Start a challenge
   */
  async startChallenge(challengeId: number, userId: number) {
    // Check if challenge exists
    const challenge = await prisma.challenge.findUnique({
      where: {
        id: challengeId,
        isActive: true,
      },
    });

    if (!challenge) {
      throw new AppError('Challenge not found', 404);
    }

    // Check if user already has this challenge
    const existing = await prisma.userChallenge.findUnique({
      where: {
        userId_challengeId: {
          userId,
          challengeId,
        },
      },
    });

    if (existing) {
      throw new AppError('Challenge already started', 400);
    }

    // Create user challenge
    const userChallenge = await prisma.userChallenge.create({
      data: {
        userId,
        challengeId,
        status: 'ACTIVE',
        currentDay: 1,
        startedAt: new Date(),
      },
      include: {
        challenge: {
          select: {
            id: true,
            title: true,
            subtitle: true,
            duration: true,
            category: true,
            goals: true,
            icon: true,
            colorTheme: true,
          },
        },
      },
    });

    return userChallenge;
  }

  /**
   * Pause a challenge
   */
  async pauseChallenge(challengeId: number, userId: number) {
    const userChallenge = await prisma.userChallenge.findUnique({
      where: {
        userId_challengeId: {
          userId,
          challengeId,
        },
      },
    });

    if (!userChallenge) {
      throw new AppError('Challenge not found', 404);
    }

    if (userChallenge.status !== 'ACTIVE') {
      throw new AppError('Challenge is not active', 400);
    }

    const updated = await prisma.userChallenge.update({
      where: {
        id: userChallenge.id,
      },
      data: {
        status: 'PAUSED',
        pausedAt: new Date(),
      },
    });

    return updated;
  }

  /**
   * Resume a paused challenge
   */
  async resumeChallenge(challengeId: number, userId: number) {
    const userChallenge = await prisma.userChallenge.findUnique({
      where: {
        userId_challengeId: {
          userId,
          challengeId,
        },
      },
    });

    if (!userChallenge) {
      throw new AppError('Challenge not found', 404);
    }

    if (userChallenge.status !== 'PAUSED') {
      throw new AppError('Challenge is not paused', 400);
    }

    const updated = await prisma.userChallenge.update({
      where: {
        id: userChallenge.id,
      },
      data: {
        status: 'ACTIVE',
        pausedAt: null,
      },
    });

    return updated;
  }

  /**
   * Complete a ritual
   */
  async completeRitual(challengeId: number, ritualId: number, userId: number) {
    // Verify the ritual belongs to the challenge
    const ritual = await prisma.ritual.findFirst({
      where: {
        id: ritualId,
        challengeId,
        isActive: true,
      },
    });

    if (!ritual) {
      throw new AppError('Ritual not found', 404);
    }

    // Get user challenge with challenge details
    const userChallenge = await prisma.userChallenge.findUnique({
      where: {
        userId_challengeId: {
          userId,
          challengeId,
        },
      },
      include: {
        challenge: {
          select: {
            duration: true,
          },
        },
      },
    });

    if (!userChallenge) {
      throw new AppError('Challenge not started', 404);
    }

    if (userChallenge.status !== 'ACTIVE') {
      throw new AppError('Challenge is not active', 400);
    }

    // Check if already completed
    const existing = await prisma.userRitualCompletion.findUnique({
      where: {
        userChallengeId_ritualId: {
          userChallengeId: userChallenge.id,
          ritualId,
        },
      },
    });

    if (existing) {
      throw new AppError('Ritual already completed', 400);
    }

    // Create completion
    await prisma.userRitualCompletion.create({
      data: {
        userId,
        userChallengeId: userChallenge.id,
        ritualId,
      },
    });

    // Get count of completed rituals for current day
    const completedToday = await prisma.userRitualCompletion.count({
      where: {
        userChallengeId: userChallenge.id,
        ritual: {
          challengeId,
          dayNumber: userChallenge.currentDay,
        },
      },
    });

    // Get total rituals for current day
    const totalToday = await prisma.ritual.count({
      where: {
        challengeId,
        dayNumber: userChallenge.currentDay,
        isActive: true,
      },
    });

    // If all rituals for today are completed, advance to next day
    if (completedToday >= totalToday) {
      const nextDay = Math.min(userChallenge.currentDay + 1, userChallenge.challenge.duration);

      await prisma.userChallenge.update({
        where: {
          id: userChallenge.id,
        },
        data: {
          currentDay: nextDay,
          ...(nextDay > userChallenge.challenge.duration && {
            status: 'COMPLETED',
            completedAt: new Date(),
          }),
        },
      });
    }

    return {
      completed: true,
      currentDay: completedToday >= totalToday ? Math.min(userChallenge.currentDay + 1, userChallenge.challenge.duration) : userChallenge.currentDay,
    };
  }

  /**
   * Get user's progress in a challenge
   */
  async getProgress(challengeId: number, userId: number) {
    const userChallenge = await prisma.userChallenge.findUnique({
      where: {
        userId_challengeId: {
          userId,
          challengeId,
        },
      },
      include: {
        challenge: {
          include: {
            rituals: {
              where: {
                isActive: true,
              },
              orderBy: {
                dayNumber: 'asc',
              },
            },
          },
        },
        completions: {
          select: {
            ritualId: true,
            completedAt: true,
          },
        },
      },
    });

    if (!userChallenge) {
      throw new AppError('Challenge not started', 404);
    }

    const completedRitualIds = new Set(userChallenge.completions.map(c => c.ritualId));

    return {
      ...userChallenge,
      completedRitualIds: Array.from(completedRitualIds),
    };
  }
}

export default new ChallengeService();
